core/src/com/unciv/logic/civilization/CivilizationInfo.kt

```diff
@@ -681,6 +687,20 @@ class CivilizationInfo {
         // Cache whether this civ gets nonstandard terrain damage for performance reasons.
         nonStandardTerrainDamage = getMatchingUniques("Units ending their turn on [] tiles take [] damage")
             .any { gameInfo.ruleSet.terrains[it.params[0]]!!.damagePerTurn != it.params[1].toInt() }
+
+        // Cache the last era each resource is used for buildings or units respectively for AI building evaluation
+        for (resource in gameInfo.ruleSet.tileResources.values.filter { it.resourceType == ResourceType.Strategic }.map { it.name }) {
```

Comment:
I doubt it's really necessary to have these values cached, it's possible this is just function inefficiency

Instead of iterating for each resource for each building and unit, you can iterate only on each building and unit sorted by era, and assign resources that you find

Reply:
Necessary no, but it makes the function a lot faster.
It makes little sense to build the whole resource structure anew for each time the function is called as for the vast majority (all?) of cases there is only a single required resource.
My first implementation was something like
```
val neededForUnits = ruleSet.units.values.any { (it.obsoleteTech == null || !civInfo.tech.isResearched(it.obsoleteTech!!))
                    && resource in it.getResourceRequirements().keys }
```
which is decently fast, but the cached values version takes only 30% as much time for the entire function.

Still the function is not that slow in absolute terms and not called _that_ often (a couple times per turn), so idk? What do you think? I'd add in the `.requiresResource()` of course.